python-package/lightgbm/dask.py

```diff
@@ -197,21 +200,47 @@ def _train(client, data, label, params, model_factory, weight=None, **kwargs):
     master_worker = next(iter(worker_map))
     worker_ncores = client.ncores()
 
-    if 'tree_learner' not in params or params['tree_learner'].lower() not in {'data', 'feature', 'voting'}:
-        logger.warning('Parameter tree_learner not set or set to incorrect value '
-                       '(%s), using "data" as default', params.get("tree_learner", None))
+    tree_learner = None
+    for tree_learner_param in _ConfigAliases.get('tree_learner'):
+        tree_learner = params.get(tree_learner_param)
+        if tree_learner is not None:
+            break
+
+    allowed_tree_learners = {
+        'data',
+        'data_parallel',
+        'feature',
+        'feature_parallel',
+        'voting',
+        'voting_parallel'
+    }
+    if tree_learner is None:
+        logger.warning('Parameter tree_learner not set. Using "data" as default"')
         params['tree_learner'] = 'data'
+    elif tree_learner.lower() not in allowed_tree_learners:
+        logger.warning('Parameter tree_learner set to %s, which is not allowed. Using "data" as default' % tree_learner)
+        params['tree_learner'] = 'data'
+
+    local_listen_port = 12400
+    for port_param in _ConfigAliases.get('local_listen_port'):
+        val = params.get(port_param)
+        if val is not None:
+            local_listen_port = val
+            break
 
     # find an open port on each worker. note that multiple workers can run
     # on the same machine, so this needs to ensure that each one gets its
     # own port
-    local_listen_port = params.get('local_listen_port', 12400)
     worker_address_to_port = _find_ports_for_workers(
         client=client,
         worker_addresses=worker_map.keys(),
         local_listen_port=local_listen_port
     )
 
+    # num_threads is set below, so remove it and all aliases of it from params
```

Comment:
Why don't we respect user value for `num_threads` parameter?

Reply:
to be clear, this is existing behavior and this PR isn't changing it. There is a feature request to relax this a bit and make it the default behavior, but allow people to control `num_threads`: #3714. I'm still unsure about whether that's a good idea.

Ok given that, here's the reasoning for the current behavior...

When you set up Dask, each worker is a process and you tell Dask how many threads that worker should be able to use. Setting `num_threads` to `worker.nthreads` ensures that LightGBM respects that. It avoids the situation where, for example, you have 2 worker processes on the same 4-core machine both running with `num_threads = 4`, which would result in LightGBM using 8 threads.

XGBoost does this overriding and does not allow such customization, for example: https://github.com/dmlc/xgboost/blob/d6d72de339e3172143b1923a780c1674f4248fcd/python-package/xgboost/dask.py#L802-L809

It's also important that we set this worker-by-worker instead of accepting a single integer, since it's possible in Dask to run worker processes on a cluster of machines that are not all the same size.

When I pick up #3714 we can have more of a discussion about this and can involve others from the Dask community.

My best guess is that the solution we'll end up is:

* if you pass `num_threads = -1` or don't include `num_threads`, you get this behavior of `num_threads = worker.nthreads`
* if you pass `num_threads` with any value other than -1, we respect that but raise a warning explaining the issues that might cause in distributed training

Reply:
OK, got it! Thank you for the super detailed response as always! 🙂 

Sorry, I didn't notice the existence of #3714.